M8/advanced-nft/nft.sol

```diff
@@ -0,0 +1,185 @@
+        _;
+    }
+
+    modifier onlyOwner() {
+        if (msg.sender != _owner) {
+            revert NotOwner();
+        }
+        _;
+    }
+
+    constructor(bytes32 _merkleRoot) ERC721("AirDropNFT", "ADN") {
+        merkleRoot = _merkleRoot;
+        _owner = msg.sender;
+        contributorCount = 1;
+    }
+
+    function addContributor(address contributor) external onlyOwner {
+        if (stage == Stages.SaleEnded) {
+            revert NoMoreContributors();
+        }
+        contributors[contributor] = true;
+        contributorCount++;
+    }
+
+    function NextStage() external onlyOwner {
+        if (stage == Stages.PendingStart) {
+            stage = Stages.WhiteList;
+        } else if (stage == Stages.WhiteList) {
+            stage = Stages.Public;
+        }
+    }
+
+    function claimWhiteList(
+        bytes32[] calldata proof,
+        uint256 index
+    ) external atStage(Stages.WhiteList) {
+        if (stage == Stages.WhiteList) {
+            // check if already claimed
+            if (BitMaps.get(_whiteList, index)) {
+                revert AlreadyClaimed();
+            }
+
+            // verify proof
+            _verifyProof(proof, index, msg.sender);
+
+            // set airdrop as claimed
+            BitMaps.setTo(_whiteList, index, true);
+        }
+
+        // mint tokens
+        _mint(msg.sender, index);
+    }
+
+    function _verifyProof(
+        bytes32[] memory proof,
+        uint256 index,
+        address addr
+    ) private view {
+        bytes32 leaf = keccak256(
+            bytes.concat(keccak256(abi.encode(addr, index)))
+        );
+
+        if (!MerkleProof.verify(proof, merkleRoot, leaf)) {
+            revert InvalidProof();
+        }
+    }
+
+    function commit(bytes32 dataHash) external payable atStage(Stages.Public) {
+        if (mintedNfts == totalSupply) {
+            stage = Stages.SaleEnded;
+            revert SaleEnded();
+        }
+        if (msg.value != 1 ether) {
+            revert InvalidMintValue();
+        }
+        commits[msg.sender].commit = dataHash;
+        commits[msg.sender].block = uint64(block.number);
+        commits[msg.sender].revealed = false;
+    }
+
+    function reveal(bytes32 revealHash) external atStage(Stages.Public) {
+        if (commits[msg.sender].revealed) {
+            revert AlreadyMinted();
+        }
+        if (uint64(block.number) < commits[msg.sender].block + 10) {
+            revert stillEarlyToMint(block.number);
+        }
+
+        if (getHash(revealHash) != commits[msg.sender].commit) {
+            revert WrongHash();
+        }
+
+        bytes32 blockHash = blockhash(commits[msg.sender].block);
+        uint256 tokenId = uint256(
+            keccak256(abi.encodePacked(blockHash, revealHash))
+        ) % (totalSupply - 10);
+        // to bypass the 10 ids of whitelabled nfts
+        tokenId = tokenId + 10;
+        ++mintedNfts;
+        emit Mint(tokenId);
+
+        _mint(msg.sender, tokenId);
```

Comment:
- [ ] there is a probability of generating same tokenId. You might want to use approach that allocates tokenId based on random number aka `revealHash`. Please refer https://github.com/1001-digital/erc721-extensions/blob/6c411ceda4c06459b435000c5bc7484f0ef47c2b/contracts/RandomlyAssigned.sol